### PR TITLE
[Prim][PIR] dropout forward sink

### PR DESCRIPTION
**Pull request by kevincheng2**
### PR types

New features
### PR changes

Others 
### Description

prim `dropout` sink c++ and add `uniform` in `primitive.h`

Comparison of Python and C++ Program：
**prim in Python**
```
{
 (%0) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"x",place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[8,16,32,64],stop_gradient:[false]} : () -> pd_op.tensor<8x16x32x64xf32>
 (%1) = "pd_op.full_int_array" () {dtype:(pd_op.DataType)int64,place:(pd_op.Place)Place(cpu),stop_gradient:[true],value:[(Int64)8,(Int64)16,(Int64)32,(Int64)64]} : () -> pd_op.tensor<4xi64>
 (%2) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)0} : () -> pd_op.tensor<1xf32>
 (%3) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)1} : () -> pd_op.tensor<1xf32>
 (%4) = "pd_op.uniform" (%1, %2, %3) {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),seed:(Int32)0,stop_gradient:[true]} : (pd_op.tensor<4xi64>, pd_op.tensor<1xf32>, pd_op.tensor<1xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%5) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)0.5} : () -> pd_op.tensor<1xf32>
 (%6) = "pd_op.greater_equal" (%4, %5) {stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<1xf32>) -> pd_op.tensor<8x16x32x64xb>
 (%7) = "pd_op.cast" (%6) {dtype:(pd_op.DataType)float32,stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xb>) -> pd_op.tensor<8x16x32x64xf32>
 (%8) = "pd_op.multiply" (%0, %7) {stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%9) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[8,16,32,64],stop_gradient:[true],value:(Float)0.5} : () -> pd_op.tensor<8x16x32x64xf32>
 (%10) = "pd_op.divide" (%8, %9) {stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%11) = "pd_op.cast" (%7) {dtype:(pd_op.DataType)uint8,stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xu8>
 (%12) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)1} : () -> pd_op.tensor<1xf32>
 (%13) = "pd_op.full_like" (%10, %12) {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<1xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%14, %15) = "pd_op.divide_grad" (%8, %9, %10, %13) {axis:(Int32)-1,stop_gradient:[false,false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>, <<NULL TYPE>>
 (%16, %17) = "pd_op.multiply_grad" (%0, %7, %14) {axis:(Int32)-1,stop_gradient:[false,false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>, <<NULL TYPE>>
 (%18) = "pd_op.fetch" (%10) {col:(Int32)0,name:"fetch0",stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%19) = "pd_op.fetch" (%16) {col:(Int32)1,name:"fetch1",stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
}
```

**prim in C++**
```
{
 (%0) = "pd_op.data" () {dtype:(pd_op.DataType)float32,name:"x",place:(pd_op.Place)Place(undefined:0),shape:(pd_op.IntArray)[8,16,32,64],stop_gradient:[false]} : () -> pd_op.tensor<8x16x32x64xf32>
 (%1) = "pd_op.full_int_array" () {dtype:(pd_op.DataType)int64,place:(pd_op.Place)Place(cpu),stop_gradient:[true],value:[(Int64)8,(Int64)16,(Int64)32,(Int64)64]} : () -> pd_op.tensor<4xi64>
 (%2) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)0} : () -> pd_op.tensor<1xf32>
 (%3) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)1} : () -> pd_op.tensor<1xf32>
 (%4) = "pd_op.uniform" (%1, %2, %3) {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),seed:(Int32)0,stop_gradient:[true]} : (pd_op.tensor<4xi64>, pd_op.tensor<1xf32>, pd_op.tensor<1xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%5) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[8,16,32,64],stop_gradient:[true],value:(Float)0.5} : () -> pd_op.tensor<8x16x32x64xf32>
 (%6) = "pd_op.greater_equal" (%4, %5) {stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xb>
 (%7) = "pd_op.cast" (%6) {dtype:(pd_op.DataType)float32,stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xb>) -> pd_op.tensor<8x16x32x64xf32>
 (%8) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[8,16,32,64],stop_gradient:[true],value:(Float)0.5} : () -> pd_op.tensor<8x16x32x64xf32>
 (%9) = "pd_op.multiply" (%0, %7) {stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%10) = "pd_op.divide" (%9, %8) {stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%11) = "pd_op.cast" (%7) {dtype:(pd_op.DataType)uint8,stop_gradient:[true]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xu8>
 (%12) = "pd_op.full" () {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(cpu),shape:(pd_op.IntArray)[1],stop_gradient:[true],value:(Float)1} : () -> pd_op.tensor<1xf32>
 (%13) = "pd_op.full_like" (%10, %12) {dtype:(pd_op.DataType)float32,place:(pd_op.Place)Place(undefined:0),stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<1xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%14, %15) = "pd_op.divide_grad" (%9, %8, %10, %13) {axis:(Int32)-1,stop_gradient:[false,false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>, <<NULL TYPE>>
 (%16, %17) = "pd_op.multiply_grad" (%0, %7, %14) {axis:(Int32)-1,stop_gradient:[false,false]} : (pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>, pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>, <<NULL TYPE>>
 (%18) = "pd_op.fetch" (%10) {col:(Int32)0,name:"fetch0",stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
 (%19) = "pd_op.fetch" (%16) {col:(Int32)1,name:"fetch1",stop_gradient:[false]} : (pd_op.tensor<8x16x32x64xf32>) -> pd_op.tensor<8x16x32x64xf32>
}
```

### CODE REVIEW

File: paddle/fluid/primitive/composite/composite.h [hunk 1]
```diff
@@ -384,6 +384,65 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_decomp(
...
+    const Tensor& x,
+    const paddle::optional<Tensor>& seed_tensor,
+    const paddle::Scalar& p,
+    bool is_test,
+    const std::string& mode,
+    const int seed,
+    bool fix_seed) {
+  auto org_dtype = x.dtype();
+  bool upscale_in_train = false;
+  if (mode.compare("upscale_in_train") == 0) {
```
Comment (cyber-pioneer) [inline paddle/fluid/primitive/composite/composite.h:398]: 直接用 == 判断吧，这里是区分了大小写的，api侧已经check过了
image

File: paddle/fluid/primitive/composite/composite.h [hunk 2]
```diff
@@ -384,6 +384,65 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_decomp(
...
+      // inference: out = input
+      return std::make_tuple(x, cast<T>(mask, phi::DataType::UINT8));
+    } else {
+      // train: out = input * mask / ( 1.0 - p )
+      if (p.to<float>() == 1.0) {
+        // Process p=1. for avoid devide zero error (x*mask/(1.0-p))
+        auto zero = full<T>(phi::vectorize(x.dims()), 0.0, org_dtype);
+        return std::make_tuple(x * zero, cast<T>(zero, phi::DataType::UINT8));
+      } else {
+        auto ans = divide<T>(x * mask, ones_p);
```
Comment (cyber-pioneer) [inline paddle/fluid/primitive/composite/composite.h:431]: divide -> /

File: paddle/fluid/primitive/composite/composite.h (hunk 1 above)
Comment (kevincheng2) [reply paddle/fluid/primitive/composite/composite.h:398]: 好的

File: paddle/fluid/primitive/composite/composite.h (hunk 2 above)
Comment (kevincheng2) [reply paddle/fluid/primitive/composite/composite.h:431]: 好的

**Review by cyber-pioneer — APPROVED** [review]

**Pull request merged**